Task complited.

ZoreslavT opened the pull request:
Also three methods getNode(), unLink() & chekindex() was added;

Veronika-Tretiak commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,186 @@
+                result = result.prev;
+            }
+            return result.value;
+        }
```
You have a method getNode. Can you use it here to make you code shorter?

Veronika-Tretiak commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,186 @@
-        return null;
+        if (index < 0 || index >= size && head != null) {
+            throw new IndexOutOfBoundsException("Invalid index " + index);
+        }
```
Don’t create repeating code.

Veronika-Tretiak commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,186 @@
-        return null;
+        if (index < 0 || index >= size && head != null) {
+            throw new IndexOutOfBoundsException("Invalid index " + index);
+        }
```
Don’t create repeating code.

Veronika-Tretiak commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,186 @@
+    }
+
+    private T unLink(Node<T> currentNode) {
+        Node<T> next = currentNode.next;
+        Node<T> prev = currentNode.prev;
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            currentNode.prev = null;
+        }
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            currentNode.next = null;
+        }
+        size--;
+        return currentNode.value;
+    }
```
Is it good practice to set the value to null?

Veronika-Tretiak commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,186 @@
     public void add(T value) {
+        currentNode = new Node<>(value);
+        Node<T> t = tail;
+        if (head == null) {
```
it is better to rely on size == 0 in this case

Veronika-Tretiak commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,186 @@
+
     @Override
     public void add(T value) {
+        currentNode = new Node<>(value);
```
You create currentNode, but you use it only one time. And create a lot of newNode

KostyaNikiforov commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,171 @@
+    private Node<T> currentNode;
+    private int size;
+
+    public MyLinkedList() {
+        this.head = null;
+        this.tail = null;
+        size = 0;
+    }
```
All of these are default values. You can remove the constructor

KostyaNikiforov commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,171 @@
+            }
+        } else {
+            for (Node<T> n = head; n != null; n = n.next) {
+                if (n.value == object || n.value != null && n.value.equals(object)) {
```
If the `object` is null and n.value is null, this expression will return true

KostyaNikiforov commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,171 @@
+    }
+
+    private void checkElementIndex(int index) {
+        if (index < 0 || index >= size && head != null) {
```
Why do you check the `head` here?

KostyaNikiforov commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,171 @@
+                result = result.next;
+            }
+            return result;
+        } else if (index == size) {
+            throw new IndexOutOfBoundsException("Invalid index " + index);
+        } else {
```
Can you use `checkElementIndex` method instead of this if-block?

KostyaNikiforov commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,171 @@
+        currentNode = new Node<>(value);
+        if (size == 0) {
+            head = currentNode;
+        } else if (tail == null) {
+            tail = currentNode;
+            head.next = tail;
+            tail.prev = head;
+        } else {
+            Node<T> lastNode = tail;
+            tail = currentNode;
+            tail.prev = lastNode;
+            tail.prev.prev.next = lastNode;
+            lastNode.next = tail;
+        }
```
If the size if not 0, it would be enough to just find the last element and set the `next` field with the new node

ZoreslavT commented:
1) Changed method add();
2) Constructor was removed
3) Deleted one of the blocks if in method remove(T object).
4) Used checkElementIndex() in method getNode(int index).

ZoreslavT commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,149 @@
+            return result;
+        } else {
+            checkElementIndex(index);
+            Node<T> result = tail;
```
Agree, i need only checkElementIndex().
